File: RealRefIce.py
import numpy as np
from modules.interpolateFreq import *
import matplotlib.pyplot as plt
import sys
import time

filePath=sys.argv[1];

T=np.linspace(10,100,51)
pi=np.pi
omega=2*pi/T
RC=np.zeros((21,4),dtype=complex)

npts=199
nev=16
omeganew=interpolateCoeffsFreq(2*pi/100,2*pi/10,omega,nev,filePath+"2_ModesMatrix/",npts,1)
LAM=buildLam(filePath)

plt.figure(figsize=[10,4])
plt.title("Modal Amplitudes")
for m in np.arange(0,6):
    L=LAM[:,m]
    plt.plot(omeganew/(2*pi),abs(L),linewidth=2,label="$|\lambda_"+str(m+1)+"|$")
plt.legend()
plt.xlabel('$\omega$')
plt.savefig(sys.argv[2]+"_1.pdf",bbox_inches='tight')

## Interpolating reflection coefficients
V=interpolateRefCoeff(omega,omeganew,nev,filePath+"2_RefCoeff/","C")
RC=buildRMat(LAM,filePath,"C")
plt.figure(figsize=[10,4])
plt.plot(omeganew/(2*pi),np.transpose(abs(RC)),linewidth=2,label="R($\omega$)")

## Interpolating Transmission coefficients
V=interpolateRefCoeff(omega,omeganew,nev,filePath+"2_RefCoeff/","T")
RT=buildRMat(LAM,filePath,"T")
plt.plot(omeganew/(2*pi),np.transpose(abs(RT)),linewidth=2,label="T($\omega$)")
plt.plot(omeganew/(2*pi),np.transpose(abs(RT)**2+abs(RC)**2),linewidth=2,label="$1.0$")
plt.title("$T(\omega)$ and $R(\omega)$")
plt.legend()
plt.xlabel('$\omega$')
plt.savefig(sys.argv[2]+"_2.pdf",bbox_inches='tight')

# Check Energy Conservation
#plt.xlim([1/80,0.2])
plt.ylim([0,1.1])
print(filePath)

#print("Sleep")
#time.sleep(10)

#Pearson Moscowitz Spectrum.
g=9.8
omega0=(g/15)
alpha=0.0081
beta=0.74
omeganew1=2*pi*np.linspace(0.01,0.45,200)
S=alpha*g**2/omeganew1**5*np.exp(-beta*(omega0/omeganew1)**4)+0.05;
plt.figure(figsize=[5,4])
plt.plot(omeganew1/(2*pi),S,label='Pearson Moskowitz Spectrum')
plt.xlabel('$\omega$ (in Hz)')
plt.ylabel('PSD')
S1=np.loadtxt('/Users/balaje/data1.txt',delimiter=',')
plt.plot(S1[:,0],S1[:,1],label='Kristensen et al.')
plt.legend()
plt.savefig(sys.argv[2]+"_5.pdf",bbox_inches= 'tight')
plt.show()



omeganew1=omeganew;
S=alpha*g**2/omeganew1**5*np.exp(-beta*(omega0/omeganew1)**4)+0.05;
S=omeganew**0


plt.figure(figsize=[5,4])
maxDisp=np.loadtxt("./dispVsFreq.dat")
plt.plot(omeganew/(2*pi),S*maxDisp[:,0],label="$u_x$")
plt.plot(omeganew/(2*pi),S*maxDisp[:,1],label="$u_y$")
plt.legend()
plt.xlabel('$\omega$')
#plt.xlim([1/100,0.45])
plt.title('Displacements')
plt.savefig(sys.argv[2]+"_3.pdf",bbox_inches='tight')


plt.figure(figsize=[5,4])
maxStrain=np.loadtxt("./strainVsFreq.dat")
plt.plot(omeganew/(2*pi),S*maxStrain[:,0],label="$e_{xx}$")
plt.plot(omeganew/(2*pi),S*maxStrain[:,1],label="$e_{xy}$")
plt.plot(omeganew/(2*pi),S*maxStrain[:,2],label="$e_{yy}$")
plt.legend()
plt.xlabel('$\omega$')
plt.title('Strain')
#plt.xlim([1/100,0.45])
plt.savefig(sys.argv[2]+"_4.pdf",bbox_inches='tight')

File: RealRefIce.m
%% Run the following command in the shell to obtain an initial set of frequency domain solutions.

% for i in $(seq 15 2.5 80); do mpirun -np 2 FreeFem++-mpi -v 0 iceberg.edp -N1 20 -N2 30 -Tr $i -L 3000 -H 2000 -h 200 -nev 8 -iter $(echo $i/2.5-5 | bc) > /dev/null; echo "Done $i"; done 

% This may take a while depending on your computer. Once the command
% finishes running, run this MATLAB script!

clc
clear

SolutionDir='TEST/';

%% From the finite element problems; (Sample Shell command to run the frequency domain solutions for iceberg.edp)
npts=27;
RC=zeros(npts,4);
for m=1:npts
    RC(m,:)=load([SolutionDir,'2_RefCoeff/rc',num2str(m),'.dat']);
end
RC1=RC(:,1)+1i*RC(:,2);
TC=RC(:,3)+1i*RC(:,4);

%% Interpolate the linear system and solve
a=15;
b=80;
T=linspace(a,b,npts);
omega=2*pi./T;
NptsNew=300;
[omegaNew,detH,conH]=interpolateFreq(2*pi/b,2*pi/a,omega,8,[SolutionDir,'2_ModesMatrix/'],NptsNew-1,1);
ReLAMBDA=load([SolutionDir,'2_ModesMatrix/Interpolated_L/lambdaRe.dat']);
ImLAMBDA=load([SolutionDir,'2_ModesMatrix/Interpolated_L/lambdaIm.dat']);
LAM=ReLAMBDA+1i*ImLAMBDA;

%% Interpolate Reflection Coeffs
V1=interpRefCoeffs(omega,omegaNew,8,[SolutionDir,'2_RefCoeff'],'C');
V2=interpRefCoeffs(omega,omegaNew,8,[SolutionDir,'2_RefCoeff'],'T');
% and load
RefCDifRe=load([SolutionDir,'2_RefCoeff/Interpolated_R/refCDifRe.dat']);
RefCDifIm=load([SolutionDir,'2_RefCoeff/Interpolated_R/refCDifIm.dat']);
RefTDifRe=load([SolutionDir,'2_RefCoeff/Interpolated_R/refTDifRe.dat']);
RefTDifIm=load([SolutionDir,'2_RefCoeff/Interpolated_R/refTDifIm.dat']);

RCDif=RefCDifRe+1i*RefCDifIm;
RTDif=RefTDifRe+1i*RefTDifIm;

RefCRadRe=load([SolutionDir,'2_RefCoeff/Interpolated_R/refCRadRe.dat']);
RefCRadIm=load([SolutionDir,'2_RefCoeff/Interpolated_R/refCRadIm.dat']);
RefTRadRe=load([SolutionDir,'2_RefCoeff/Interpolated_R/refTRadRe.dat']);
RefTRadIm=load([SolutionDir,'2_RefCoeff/Interpolated_R/refTRadIm.dat']);

RCRad=RefCRadRe+1i*RefCRadIm;
RTRad=RefTRadRe+1i*RefTRadIm;

RC=zeros(NptsNew,1);
RT=zeros(NptsNew,1);

%% Reconstruct RefCoeffs
for indx=1:NptsNew
    L=LAM(indx,:);
    RC(indx)=RCDif(indx)+dot(conj(RCRad(indx,:)),L);
    RT(indx)=RTDif(indx)+dot(conj(RTRad(indx,:)),L);
end

fig=figure(1);
set(fig,'Position',[273   123   967   582]);
subplot(3,1,1);
plot(omega,abs(RC1),'+','linewidth',2);
hold on
plot(omegaNew,abs(RC),'linewidth',2)
xlabel('$\omega$');
ylabel('$R(\omega)$');
legend('Coarse $\omega$ Space','Fine $\omega$ Space')
xlim([omegaNew(1),omegaNew(end)])

subplot(3,1,2);
plot(omegaNew,sqrt(abs(RT).^2+abs(RC).^2),'k','linewidth',2);
hold on
plot(omegaNew,abs(RT),'r','linewidth',2);
plot(omegaNew,abs(RC),'b','linewidth',2);
xlabel('$\omega$');
legend('1','$T(\omega)$','$R(\omega)$');
xlim([omegaNew(1),omegaNew(end)])

subplot(3,1,3);
for indx=1:4
    L=LAM(:,indx);
    plot(omegaNew,abs(L),'DisplayName',['$|\lambda_{',num2str(indx),'}|$'],'linewidth',2);    
    hold on
end
legend show
title('Coefficients vs $\omega$');
xlabel('$\omega$');
ylabel('$|\lambda_j|$');
xlim([omegaNew(1),omegaNew(end)])
%ylim([0,0.8])

%% Function
function V=interpRefCoeffs(omega,omegaNew,Nev,filePath,TorC)
rd = zeros(length(omega),1); %Diffraction RefCoeffs
rr = zeros(length(omega),Nev); %Radiation RefCoeffs
for m=1:length(omega)
    rcDiff = load([filePath,'/RefCoeff_Dif/ref',TorC,num2str(m),'.dat']);
    rcRad = load([filePath,'/RefCoeff_Rad/ref',TorC,num2str(m),'.dat']);
    
    rd(m) = rcDiff(1)+1i*rcDiff(2);
    rr(m,:) = (rcRad(:,1)+1i*rcRad(:,2)).';
end
rdNewRe = interp1(omega,real(rd),omegaNew,'pchip');
rdNewIm = interp1(omega,imag(rd),omegaNew,'pchip');
% Loop over each mode and interpolate the reflection coefficients
rcNew = zeros(length(omegaNew),Nev);
for m=1:Nev
    rccRe = interp1(omega,real(rr(:,m)),omegaNew,'pchip');
    rccIm = interp1(omega,imag(rr(:,m)),omegaNew,'pchip');
    rcc=rccRe+1i*rccIm;
    rcNew(:,m) = rcc;
end

%% Write the interpolated reflection coefficients to a file.
dlmwrite([filePath,'/Interpolated_R/ref',TorC,'DifRe.dat'],rdNewRe);
dlmwrite([filePath,'/Interpolated_R/ref',TorC,'DifIm.dat'],rdNewIm);
dlmwrite([filePath,'/Interpolated_R/ref',TorC,'RadRe.dat'],real(rcNew));
dlmwrite([filePath,'/Interpolated_R/ref',TorC,'RadIm.dat'],imag(rcNew));

V=0;
end
